Signal/src/ViewControllers/CropScaleImageViewController.swift

```diff
@@ -174,52 +175,64 @@ class CropScaleImageViewController: OWSViewController {
-                                                                target:self,
-                                                                action:#selector(cancelPressed))
-        self.navigationItem.title = NSLocalizedString("CROP_SCALE_IMAGE_VIEW_TITLE",
-                                                      comment: "Title for the 'crop/scale image' dialog.")
-
         createViews()
     }
 
     // MARK: - Create Views
 
     private func createViews() {
-        let previewTopMargin: CGFloat = 30
-        let previewHMargin: CGFloat = 20
 
         let contentView = UIView()
+        contentView.backgroundColor = UIColor.black
         self.view.addSubview(contentView)
-        contentView.autoPinWidthToSuperview(withMargin:previewHMargin)
-        contentView.autoPin(toTopLayoutGuideOf: self, withInset:previewTopMargin)
+        contentView.autoPinEdgesToSuperviewEdges()
 
-        createButtonRow(contentView:contentView)
-
-        let imageHMargin: CGFloat = 0
         let imageView = OWSLayerView(frame:CGRect.zero, layoutCallback: {[weak self] _ in
             guard let strongSelf = self else { return }
             strongSelf.updateImageLayout()
         })
         imageView.clipsToBounds = true
         self.imageView = imageView
         contentView.addSubview(imageView)
-        imageView.autoPinWidthToSuperview(withMargin:imageHMargin)
-        imageView.autoVCenterInSuperview()
-        imageView.autoPinToSquareAspectRatio()
+        imageView.autoPinEdgesToSuperviewEdges()
 
         let imageLayer = CALayer()
         self.imageLayer = imageLayer
         imageLayer.contents = srcImage.cgImage
         imageView.layer.addSublayer(imageLayer)
 
-        let dashedBorderLayer = CAShapeLayer()
-        self.dashedBorderLayer = dashedBorderLayer
-        dashedBorderLayer.strokeColor = UIColor.ows_materialBlue().cgColor
-        dashedBorderLayer.lineDashPattern = [10, 10]
-        dashedBorderLayer.lineWidth = 4
-        dashedBorderLayer.fillColor = nil
-        imageView.layer.addSublayer(dashedBorderLayer)
+        let maskingView = OWSBezierPathView()
+        contentView.addSubview(maskingView)
+
+        maskingView.configureShapeLayerBlock = { layer, bounds in
+            let path = UIBezierPath(rect: bounds)
+
+            let radius = min(bounds.size.width, bounds.size.height) * 0.5 - self.maskMargin
+            // Center the circle's bounding rectangle
+            let circleRect = CGRect(x: bounds.size.width * 0.5 - radius, y: bounds.size.height * 0.5 - radius, width: radius * 2, height: radius * 2)
+            let circlePath = UIBezierPath(roundedRect: circleRect, cornerRadius: radius)
+            path.append(circlePath)
+            path.usesEvenOddFillRule = true
+
+            layer.path = path.cgPath
+            layer.fillRule = kCAFillRuleEvenOdd
+            layer.fillColor = UIColor.black.cgColor
+            layer.opacity = 0.7
+        }
+        maskingView.autoPinEdgesToSuperviewEdges()
```

Comment:
This is a big change, if I'm reading it properly.  I'm pretty sure this changes the coordinate system of the crop state and will affect the output of `generateDstImage()` so that this is no longer WYSIWIG.  I did a little testing and it looks close to but a little off.  

To Repro:

* Take a screenshot of a conversation with a lot of messages - conversation view should have a lot of features with sharp edges, like message bubbles.
* Crop to the edges of a message bubble that is near a _corner_ of the screen, so that the edges of the bubble are near but inside the edges of the crop region.  
* Exit view. Look at the cropped output avatar.  The crop will be off.

### Before Adjusting Crop

![screen shot 2017-09-01 at 3 17 05 pm](https://user-images.githubusercontent.com/625803/29984521-d2f03066-8f28-11e7-995c-b2eb129d16e2.png)

### After Adjusting Crop

![screen shot 2017-09-01 at 3 17 18 pm](https://user-images.githubusercontent.com/625803/29984519-d2e88f1e-8f28-11e7-8dd6-8219fe2db94d.png)

### Cropped output

![screen shot 2017-09-01 at 3 17 24 pm](https://user-images.githubusercontent.com/625803/29984520-d2e94a8a-8f28-11e7-986a-c618018d9a37.png)

Might I suggest instead:

* Keep the current coordinate system, where the crop region is represented as a translation and scale relative to source image coordinates, using `srcTranslation` and `imageScale` properties.
* Modify maskingView's bounds to correspond to just the crop circle's frame.  
* Use separate OWSBezierPathView to overlay the rest of the background image with the semi-transparent dark overlay.

Reply:
If you don't want to change the maskingView's layout, let me know - I can take a crack at fixing the image rendering so that the crop is WYSIWYG.